src/main/scripts/src/app/shared/permissions/table/permissions-table.component.ts

```diff
@@ -47,54 +52,342 @@ export class PermissionsTableComponent implements OnInit {
+    }
+  }
+
+  public updateCheckedRoles(role: string, event) {
+    this.rolesCheckbox[role] = event.target.checked;
+  }
+
+  public addOrganizationUsersGroupsRole(orgCode: string, role: string, users: string[], groups: string[]) {
+    this.rolesService.addOrganizationUsersGroupsRole(orgCode, role, users, groups)
+      .subscribe();
+  }
+
+  public removeOrganizationUsersGroupsRole(orgCode: string, role: string, users: string[], groups: string[]) {
+    this.rolesService.removeOrganizationUsersGroupsRole(orgCode, role, users, groups)
+      .subscribe();
+  }
+
+  public addProjectUsersGroupsRole(orgCode: string, projectCode: string, role: string, users: string[], groups: string[]) {
+    this.rolesService.addProjectUsersGroupsRole(orgCode, projectCode, role, users, groups)
+      .subscribe();
+  }
+
+  public removeProjectUsersGroupsRole(orgCode: string, projectCode: string, role: string, users: string[], groups: string[]) {
+    this.rolesService.removeProjectUsersGroupsRole(orgCode, projectCode, role, users, groups)
+      .subscribe();
+  }
+
+  private addOrganizationRoles(addedRoles: string[], name: string) {
+    this.route.paramMap.subscribe((params: ParamMap) => {
+      let orgCode: string = params.get('organizationCode');
+      switch (this.entityType) {
+        case 'users':
+          for (let roleIdx in this.roles) {
+            if (this.rolesCheckbox[this.roles[roleIdx]] && this.roles.hasOwnProperty(roleIdx)) {
+              this.addOrganizationUsersGroupsRole(orgCode, this.roles[roleIdx], [name], []);
+              addedRoles.push(this.roles[roleIdx]);
+            }
+          }
+        break;
+        case 'groups':
+          for (let roleIdx in this.roles) {
+            if (this.rolesCheckbox[this.roles[roleIdx]] && this.roles.hasOwnProperty(roleIdx)) {
+              this.addOrganizationUsersGroupsRole(orgCode, this.roles[roleIdx], [], [name]);
+              addedRoles.push(this.roles[roleIdx]);
+            }
+          }
+          break;
+      }
+    });
+  }
+
+  private addProjectRoles(addedRoles: string[], name: string) {
+    this.route.paramMap.subscribe((params: ParamMap) => {
+      let orgCode: string = params.get('organizationCode');
+      let proCode: string = params.get('projectCode');
+      switch (this.entityType) {
+        case 'users':
+          for (let roleIdx in this.roles) {
+            if (this.rolesCheckbox[this.roles[roleIdx]] && this.roles.hasOwnProperty(roleIdx)) {
+              this.addProjectUsersGroupsRole(orgCode, proCode, this.roles[roleIdx], [name], []);
+              addedRoles.push(this.roles[roleIdx]);
+            }
+          }
+        break;
+        case 'groups':
+          for (let roleIdx in this.roles) {
+            if (this.rolesCheckbox[this.roles[roleIdx]] && this.roles.hasOwnProperty(roleIdx)) {
+              this.addProjectUsersGroupsRole(orgCode, proCode, this.roles[roleIdx], [], [name]);
+              addedRoles.push(this.roles[roleIdx]);
```

Comment:
The same block of code is repeated four times above. It can be extracted into one new method if properly parametrized.